File: web/src/lib/schemas/session.ts
import { z } from 'zod';

export const sessionSummarySchema = z.object({
  id: z.string(),
  source: z.string(),
  title: z.string(),
  firstUserMessage: z.string().nullable().optional(),
  cwd: z.string().nullable().optional(),
  cwdShort: z.string().nullable().optional(),
  messageCount: z.number(),
  tokensTotal: z.number().nullable().optional(),
  model: z.string().nullable().optional(),
  lastActivity: z.string().nullable().optional(),
  startedAt: z.string().nullable().optional(),
  hidden: z.boolean().optional(),
  isRunning: z.boolean().optional(),
});

export const messageSchema = z.object({
  id: z.string(),
  role: z.enum(['user', 'assistant', 'tool_use', 'tool_result', 'meta', 'system']),
  text: z.string().nullable().optional(),
  toolName: z.string().nullable().optional(),
  toolInput: z.any().nullable().optional(),
  toolOutput: z.any().nullable().optional(),
  toolUseId: z.string().nullable().optional(),
  toolStatus: z.string().nullable().optional(),
  timestamp: z.string().nullable().optional(),
  model: z.string().nullable().optional(),
  images: z.array(z.any()).optional(),
});

export const sessionDetailSchema = z.object({
  summary: sessionSummarySchema,
  messages: z.array(messageSchema),
  subagents: z.array(z.any()).optional(),
  resumeHints: z.any().nullable().optional(),
});

File: src-server/src/models/session.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
#[serde(rename_all = "camelCase")]
#[allow(dead_code)]
pub struct SessionSummary {
    pub id: String,
    pub source: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub first_user_message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd_short: Option<String>,
    pub message_count: u32,
    #[ts(type = "number | null")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tokens_total: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_activity: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub started_at: Option<String>,
    #[serde(default)]
    pub hidden: bool,
    #[serde(default)]
    pub is_running: bool,
}
